core/evm.go

```diff
@@ -501,3 +527,183 @@ func Transfer(db vm.StateDB, sender, recipient common.Address, amount *big.Int,
+					return nil, err
+				}
+				return common.LeftPadBytes(
+					wrapper.TotalDelegation().Bytes(), 32,
+				), nil
+			}
+		case "BalanceAvailableForRedelegation":
+			{
+				delegations, err := chain.ReadDelegationsByDelegatorAt(
+					roStakeMsg.DelegatorAddress,
+					big.NewInt(0).Sub(ref.Number(), big.NewInt(1)),
+				)
+				if err != nil {
+					return nil, err
+				}
+				redelegationTotal := big.NewInt(0)
+				for _, delegationIndex := range delegations {
+					wrapper, err := db.ValidatorWrapper(delegationIndex.ValidatorAddress, true, false)
+					if err != nil {
+						return nil, err
+					}
+					if uint64(len(wrapper.Delegations)) > delegationIndex.Index {
+						delegation := wrapper.Delegations[delegationIndex.Index]
+						for _, undelegation := range delegation.Undelegations {
+							if undelegation.Epoch.Cmp(ref.Epoch()) < 1 { // Undelegation.Epoch < currentEpoch
+								redelegationTotal.Add(redelegationTotal, undelegation.Amount)
+							}
+						}
+					}
+				}
+				return common.LeftPadBytes(
+					redelegationTotal.Bytes(), 32,
+				), nil
+			}
+		case "ValidatorCommissionRate":
+			{
+				wrapper, err := db.ValidatorWrapper(roStakeMsg.ValidatorAddress, true, false)
+				if err != nil {
+					return nil, err
+				}
+				// Solidity doesn't support float directly, so just return the bytes
+				// the bytes can't be an exact number but can be used for comparison
+				// for example, a 1.1 commission rate shows up as 1100000000000000000
+				// divide by 1e18 (in Python) to get the actual rate 1.1
+				return common.LeftPadBytes(
+					wrapper.CommissionRates.Rate.Bytes(), 32,
+				), nil
+			}
+		case "ValidatorStatus":
+			{
+				wrapper, err := db.ValidatorWrapper(roStakeMsg.ValidatorAddress, true, false)
+				if err != nil {
+					return nil, err
+				}
+				return wrapper.Status.Bytes(), nil
+			}
+		case "BalanceDelegatedByDelegator":
+			{
+				delegations, err := chain.ReadDelegationsByDelegator(roStakeMsg.DelegatorAddress)
+				if err != nil {
+					return nil, err
+				}
+				answer := big.NewInt(0)
+				for _, delegation := range delegations {
+					wrapper, err := db.ValidatorWrapper(delegation.ValidatorAddress, true, false)
+					if err != nil {
+						return nil, err
+					}
+					answer = answer.Add(answer, wrapper.Delegations[delegation.Index].Amount)
+				}
+				return common.LeftPadBytes(answer.Bytes(), 32), nil
+			}
+		case "MedianRawStakeSnapshot":
+			{
+				epoch := big.NewInt(0).Add(ref.Epoch(), big.NewInt(1))
+				instance := shard.Schedule.InstanceForEpoch(epoch)
+				res, err := committee.NewEPoSRound(epoch, chain, chain.Config().IsEPoSBound35(epoch), instance.SlotsLimit(), int(instance.NumShards()))
+				if err != nil {
+					return nil, err
+				}
+				return res.MedianStake.Bytes(), nil
+			}
+		case "TotalStakingSnapshot":
+			{
+				candidates := chain.ValidatorCandidates()
+				fmt.Println(len(candidates))
+				staked := big.NewInt(0)
+				for i := range candidates {
```

Comment:
This is heavy too. Reading all validator wrapper data is huge. Each validator wrapper contains all the delegator info. This is basically reading everything about staking data. We shouldn't calculate on the fly and expose any aggregate info unless it's pre-calculated.